internal/services/notification/runwebhookdelivery.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2023 Sorint.lab
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package notification
+
+import (
+	"bytes"
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/sorintlab/errors"
+
+	"agola.io/agola/internal/sqlg/lock"
+	"agola.io/agola/internal/sqlg/sql"
+	"agola.io/agola/internal/util"
+	"agola.io/agola/services/notification/types"
+)
+
+const (
+	MaxRunWebhookDeliveriesLimit = 40
+	RunWebhookDeliveriesLockKey  = "runwebhookdeliveryevents"
+
+	// runWebhookDeliveriesInterval is the maximum time to wait for run webhook delivery timeout handler
+	runWebhookDeliveriesInterval = time.Second * 1
+)
+
+func (n *NotificationService) RunWebhookDeliveriesHandlerLoop(ctx context.Context) {
+	for {
+		if err := n.runWebhookDeliveriesHandler(ctx); err != nil {
+			n.log.Err(err).Send()
+		}
+
+		sleepCh := time.NewTimer(runWebhookDeliveriesInterval).C
+		select {
+		case <-ctx.Done():
+			return
+		case <-sleepCh:
+		}
+	}
+}
+
+func (n *NotificationService) runWebhookDeliveriesHandler(ctx context.Context) error {
+	runWebhookDeliveryIDs := make([]string, 0)
+	err := n.d.Do(ctx, func(tx *sql.Tx) error {
```

Comment:
In this case it will work anyway since you're calling  runWebhookDeliveriesHandler every second. BUT the right solution will be to handle all the unprocessed runWebhookDeliveries using multiple limited queries and starting from the last processed sequence like done already in other places:
 
https://github.com/agola-io/agola/blob/master/internal/services/runservice/api/api.go#L1002

Reply:
@sgotti do you think is better to save the last sequence processed(in a new table) in this pr ? or the runWebhookDeliveriesHandler must start from the last delivery as done for runEvents?

Reply:
You don't have tosave the last sequence in a table. You already have the logic to start from the last not processed delivery. Just add an additional inner loop to process in batches all the not processed deliveries. See the already existing code.